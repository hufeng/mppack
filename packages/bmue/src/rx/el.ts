import { TPath, TRxLang, TRxLangHandler } from '../types';

export class ELang {
  constructor(name: string, lang: TRxLang) {
    this.name = name;
    this.lang = lang;
  }

  name: string;
  lang: TRxLang;

  parse() {
    const lang = this.lang.slice();
    const handler = lang.pop() as TRxLangHandler;
    const deps = lang as Array<TPath>;
    return {
      name: this.name,
      deps,
      handler
    };
  }
}

//Factory method
export const EL = (name: string, lang: TRxLang) => new ELang(name, lang);
