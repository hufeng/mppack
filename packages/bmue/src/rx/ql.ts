import { TPath, TQueryLang, TQueryLangHandler } from '../types';

let uuid = 0;

export class QueryLang {
  constructor(name: string, lang: TQueryLang) {
    this.id = ++uuid;
    this.name = name;
    this.lang = lang;
  }

  id: number;
  name: string;
  lang: TQueryLang;

  parse() {
    const lang = this.lang.slice();
    const handler = lang.pop() as TQueryLangHandler;
    const deps = lang as Array<TPath>;
    return {
      id: this.id,
      name: this.name,
      deps,
      handler
    };
  }
}

//Factory method
export const QL = (name: string, lang: TQueryLang) => new QueryLang(name, lang);
