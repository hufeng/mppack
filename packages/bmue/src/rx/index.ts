import isEqual from 'lodash.isequal';
import { IRxParams, TRxLangHandler } from '../types';
import { getPathVal } from '../util';
import { ELang } from './el';
import { QueryLang } from './ql';

/**
 * design a simple reactive class
 * from bigQuery, we can computed QL
 * from computeEL, we can effect our side-effect
 */
export default class Rx {
  constructor(params: IRxParams) {
    const { dev = false, data = {}, getter = {}, effect = {} } = params;
    this.dev = dev;
    this.getter = getter;
    this.initData(data);
    this.effect = this.processEL(effect);
  }

  //merge all data
  data = {};
  //current status, dev can give more log
  dev = false;
  //record all side effect
  effect: Array<TRxLangHandler> = [];
  //record all ql
  getter: { [key: string]: QueryLang } = Object.create({});
  //cache ql result
  _cacheQL = {} as { [key: number]: Array<any> };

  /**
   * compute ql's value and merge to this.data
   */
  initData = (data: Object) => {
    const rxData = this.computeQL(data);
    this.data = {
      ...data,
      rx: {
        ...rxData
      }
    };
    if (this.dev) {
      console.log('初始化data:', this.data);
    }
  };

  computeQL = (data: Object) => {
    let rx = {};
    //compute each ql
    for (let qlName in this.getter) {
      if (this.getter.hasOwnProperty(qlName)) {
        const ql = this.getter[qlName];
        const value = this.bigQuery(data, ql);
        if (!isEqual(value, (data as any).rx[qlName])) {
          rx[qlName] = value;
        }
      }
    }
    return rx;
  };

  processEL = (effect = {}) => {
    const effects = [];
    Object.keys(effect).forEach(key => {
      effects.push(this.parseEL(this.data, effect[key]));
    });
    return effects;
  };

  bigQuery(data: Object, ql: QueryLang) {
    const { id, name, handler, deps } = ql.parse();

    if (this.dev) {
      console.groupCollapsed(
        `=================ql#${name}=====================`
      );
    }

    //initialize querylang cache
    !this._cacheQL[id] && (this._cacheQL[id] = []);

    //get path value
    let isChanged = false;
    deps.forEach((dep, i) => {
      //suport ql nested ql
      let val = null;
      if (dep instanceof QueryLang) {
        val = this.bigQuery(data, dep);
      } else {
        val = getPathVal(data, dep);
      }
      if (!isEqual(val, this._cacheQL[id][i])) {
        isChanged = true;
        this._cacheQL[id][i] = val;
      }
    });

    if (isChanged) {
      const len = deps.length;
      const pathVals = this._cacheQL[id].slice(0, len);
      const result = handler(...pathVals);
      this._cacheQL[id][len + 1] = result;

      if (this.dev) {
        console.log('update:🔥', result);
        console.groupEnd();
      }
      return result;
    } else {
      const len = deps.length;
      const result = this._cacheQL[id][len + 1];

      if (this.dev) {
        console.log('cache:😆', result);
        console.groupEnd();
      }
      return result;
    }
  }

  parseEL = (data: Object, el: ELang) => {
    const cache = [];
    const { name, deps, handler } = el.parse();

    for (let dep of deps) {
      cache.push(
        dep instanceof QueryLang
          ? this.bigQuery(data, dep)
          : getPathVal(data, dep)
      );
    }

    return (newData: Object, ctx: Object) => {
      if (this.dev) {
        console.groupCollapsed(
          `=================el#${name}=====================`
        );
      }
      let isChanged = false;
      deps.forEach((dep, i) => {
        let pathVal =
          dep instanceof QueryLang
            ? this.bigQuery(newData, dep)
            : getPathVal(newData, dep);
        if (!isEqual(pathVal, cache[i])) {
          cache[i] = pathVal;
          isChanged = true;
        }
      });

      if (isChanged) {
        if (this.dev) {
          console.log(`${name}#changed:🔥`);
          console.groupEnd();
        }
        handler.apply(ctx, [...cache, ctx]);
      } else {
        if (this.dev) {
          console.groupEnd();
        }
      }
    };
  };
}
