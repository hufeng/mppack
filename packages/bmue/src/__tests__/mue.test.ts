import Action from '../action';
import Mue from '../mue';
import { EL } from '../rx/el';
import { QL } from '../rx/ql';

const helloRL = EL('helloRL', [
  //data path
  ['list', 0, 'name'],
  name => console.log(name)
]);

const hello = QL('helloQL', [
  // data path
  ['list', 0],
  ({ name, mott }) => `${name}-${mott}`
]);

//mock miniapp Page object
window['Page'] = jest.fn(obj => obj);

it('test mue', () => {
  const PaginationMixin = {
    data: {
      pageList: []
    },
    pageNum: 0,
    pageSize: 20,
    onPullDownRefresh() {
      // this.fetchData()
    },
    onReachBottom() {
      // this.fetchData()
    },
    fetchData(url, params) {},
    refreshFetchData(url, paramm) {}
  };

  const action = Action({
    sayWorld() {}
  });

  const mue = Mue({
    getter: {
      hello
    },
    effect: { helloRL },
    mixins: [PaginationMixin],
    data: {
      list: [{ id: 1, name: 'mue', mott: 'easy miniapp' }]
    },
    action,
    sayHello() {}
  });

  expect(mue).toMatchSnapshot();
});
