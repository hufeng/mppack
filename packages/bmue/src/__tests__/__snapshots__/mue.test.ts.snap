// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test mue 1`] = `
Object {
  "_cacheQL": Object {
    "1": Array [
      Object {
        "id": 1,
        "mott": "easy miniapp",
        "name": "mue",
      },
      undefined,
      "mue-easy miniapp",
    ],
  },
  "computeQL": [Function],
  "data": Object {
    "list": Array [
      Object {
        "id": 1,
        "mott": "easy miniapp",
        "name": "mue",
      },
    ],
    "pageList": Array [],
    "rx": Object {
      "hello": "mue-easy miniapp",
    },
  },
  "dev": false,
  "effect": Array [
    [Function],
  ],
  "fetchData": [Function],
  "getter": Object {
    "hello": QueryLang {
      "id": 1,
      "lang": Array [
        Array [
          "list",
          0,
        ],
        [Function],
      ],
      "name": "helloQL",
    },
  },
  "initData": [Function],
  "onPullDownRefresh": [Function],
  "onReachBottom": [Function],
  "pageNum": 0,
  "pageSize": 20,
  "parseEL": [Function],
  "processEL": [Function],
  "refreshFetchData": [Function],
  "sayHello": [Function],
  "sayWorld": [Function],
  "setState": [Function],
  "spliceState": [Function],
}
`;
